docs/development/design/builder-maintainability.rst

```diff
@@ -0,0 +1,308 @@
+Builder Maintainability
+=======================
+
+.. contents::
+   :local:
+   :depth: 2
+
+This document extends the proposal done by Manuel at `Future Builder`_ to replace the last 3 goals:
+
+* Allow us to add a feature with a defined contract without worry about breaking old builds
+* Introduce ``build.builder: 2`` config (does not install pre-defined packages) for these new features
+* Motivate users to migrate to ``v2`` to finally deprecate this magic by educating users
+
+
+It proposes a way to version the current "magical" builder so we can freeze its current status and
+extend it if needed by doing a new release of it.
+Besides, it removes the needing of using ``build.builder: 2`` since the *new builder* will extend the "magical" builder behavior
+and just override what's required.
```

Comment:
I thought I had this idea clear in my mind when I started writing this, but then I realized that I didn't have a good and clear way as a user to disable _all the magic_. This may be a good topic to cover in our meeting.